numpy/linalg/umath_linalg.c.src

```diff
@@ -3206,42 +3245,84 @@ release_@lapack_func@(GELSD_PARAMS_t* params)
     memset(params, 0, sizeof(*params));
 }
 
+/** Compute the squared l2 norm of a contiguous vector */
+static @basetyp@
+@TYPE@_abs2(@typ@ *p, npy_intp n) {
+    npy_intp i;
+    @basetyp@ res = 0;
+    for (i = 0; i < n; i++) {
+        @typ@ el = p[i];
+#if @cmplx@
+        res += el.real*el.real + el.imag*el.imag;
+#else
+        res += el*el;
+#endif
+    }
+    return res;
+}
+
 static void
 @TYPE@_lstsq(char **args, npy_intp *dimensions, npy_intp *steps,
              void *NPY_UNUSED(func))
 {
     GELSD_PARAMS_t params;
     int error_occurred = get_fp_invalid_and_clear();
     fortran_int n, m, nrhs;
-    INIT_OUTER_LOOP_6
+    fortran_int excess;
+
+    INIT_OUTER_LOOP_7
 
     m = (fortran_int)dimensions[0];
     n = (fortran_int)dimensions[1];
     nrhs = (fortran_int)dimensions[2];
+    excess = m - n;
 
     if (init_@lapack_func@(&params, m, n, nrhs)) {
-        LINEARIZE_DATA_t a_in, b_in, x_out, s_out;
+        LINEARIZE_DATA_t a_in, b_in, x_out, s_out, r_out;
 
         init_linearize_data(&a_in, n, m, steps[1], steps[0]);
         init_linearize_data_ex(&b_in, nrhs, m, steps[3], steps[2], fortran_int_max(n, m));
-        init_linearize_data(&x_out, nrhs, fortran_int_max(n, m), steps[5], steps[4]);
-        init_linearize_data(&s_out, 1, fortran_int_min(n, m), 1, steps[6]);
+        init_linearize_data_ex(&x_out, nrhs, n, steps[5], steps[4], fortran_int_max(n, m));
+        init_linearize_data(&r_out, 1, nrhs, 1, steps[6]);
+        init_linearize_data(&s_out, 1, fortran_int_min(n, m), 1, steps[7]);
 
-        BEGIN_OUTER_LOOP_6
+        BEGIN_OUTER_LOOP_7
             int not_ok;
             linearize_@TYPE@_matrix(params.A, args[0], &a_in);
             linearize_@TYPE@_matrix(params.B, args[1], &b_in);
             params.RCOND = args[2];
             not_ok = call_@lapack_func@(&params);
             if (!not_ok) {
                 delinearize_@TYPE@_matrix(args[3], params.B, &x_out);
-                *(npy_int*) args[4] = params.RANK;
-                delinearize_@REALTYPE@_matrix(args[5], params.S, &s_out);
+                *(npy_int*) args[5] = params.RANK;
+                delinearize_@REALTYPE@_matrix(args[6], params.S, &s_out);
+
+                /* Note that linalg.lstsq discards this when excess == 0 */
+                if (excess >= 0 && params.RANK == n) {
+                    /* Compute the residuals as the square sum of each column */
+                    int i;
+                    char *resid = args[4];
+                    @ftyp@ *components = (@ftyp@ *)params.B + n;
+                    for (i = 0; i < nrhs; i++) {
+                        @ftyp@ *vector = components + i*m;
+                        /* Numpy and fortran floating types are the same size,
+                         * so this case is safe */
+                        @basetyp@ abs2 = @TYPE@_abs2((@typ@ *)vector, excess);
+                        memcpy(
```

Comment:
My poor knowledge of C does not really help here, but `memcpy` feels very odd. Can one not just assign?

Reply:
If `resid + i*r_out.column_strides` is not aligned, there's no guarantee that the write will succeed. Unless there's some gufunc magic wrapping us that ensures this is the case, `memcpy` is safer.

Also, `delinearize_@TYPE@_matrix` uses `memcpy`, which is where we copy between numpy and the other lapack buffers - so `memcpy` is at least consistent.

Also, something-something-strict-aliasing.

Reply:
A simple example: This will not work on some platforms:

```
char bytes[8] = {0};
(uint32_t *)&bytes[1] = 0x12345678;  // unaligned 32-bit write may fail
assert((uint32_t *)&bytes[0] == 0x00123456);  // assuming big endian
```

But using `memcpy` would work every time

Reply:
Ah, thanks that makes it clearer (link is a good one too)